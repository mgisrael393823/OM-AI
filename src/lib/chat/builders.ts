export { chatCompletion as buildChatCompletion, responses as buildResponses } from '@/lib/services/openai/builders'
