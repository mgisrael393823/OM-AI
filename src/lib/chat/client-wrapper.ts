export { callOpenAIWithFallback, handleStream, openai } from '@/lib/services/openai/client-wrapper'
