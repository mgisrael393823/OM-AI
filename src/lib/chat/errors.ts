import type { NextApiRequest, NextApiResponse } from 'next'

// Check if structured error format is requested
function shouldUseStructuredFormat(req: NextApiRequest): boolean {
  // Check header first
  const errorFormatHeader = req.headers['x-error-format'] as string
  if (errorFormatHeader === 'structured') {
    return true
  }
  
  // Check query parameter
  const errorFormatQuery = req.query.errorFormat as string
  if (errorFormatQuery === 'structured') {
    return true
  }
  
  return false
}

export function jsonError(
  res: NextApiResponse,
  status: number,
  code: string,
  message: string,
  requestId: string,
  req?: NextApiRequest
) {
  if (res.headersSent) return
  
  // Standard headers
  res.setHeader('Content-Type', 'application/json; charset=utf-8')
  res.setHeader('X-Request-ID', requestId)
  
  // CORS headers - echo origin if available
  const origin = req?.headers.origin || '*'
  res.setHeader('Access-Control-Allow-Origin', origin)
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, X-Request-ID, X-Correlation-ID, Authorization')
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS')
  res.setHeader('Vary', 'Origin')
  
  // Method-specific headers
  if (status === 405) {
    res.setHeader('Allow', 'POST, OPTIONS')
  }
  
  // Default to legacy format for backward compatibility
  if (!req || !shouldUseStructuredFormat(req)) {
    // Legacy format
    return res.status(status).json({
      error: message,
      code,
      requestId
    })
  }
  
  // Structured format (opt-in only)
  return res.status(status).json({
    error: {
      type: 'api_error',
      code,
      message
    },
    requestId
  })
}