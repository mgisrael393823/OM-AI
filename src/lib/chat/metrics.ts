// Basic metrics implementation - can be enhanced with real telemetry backend
export const metrics = {
  count: (name: string, value: number, tags?: Record<string, string>) => {
    // For now, just log to console in development
    if (process.env.NODE_ENV === 'development') {
      console.log(`[METRIC] ${name}: ${value}`, tags)
    }
    // TODO: Integrate with chosen telemetry backend (DataDog, New Relic, etc.)
  },
  timing: (name: string, value: number, tags?: Record<string, string>) => {
    // For now, just log to console in development
    if (process.env.NODE_ENV === 'development') {
      console.log(`[TIMING] ${name}: ${value}ms`, tags)
    }
    // TODO: Integrate with chosen telemetry backend
  }
}