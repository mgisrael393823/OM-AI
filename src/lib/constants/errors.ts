/**
 * Centralized Error Codes for OM-AI Platform
 * 
 * IMPORTANT: All error codes must be used consistently across the entire project.
 * When adding new codes, update this enum and corresponding message mappings.
 * Replace all inline error strings throughout the codebase with these enum values.
 */

export enum ERROR_CODES {
  // Authentication & Authorization
  MISSING_TOKEN = 'MISSING_TOKEN',
  INVALID_TOKEN = 'INVALID_TOKEN',
  UNAUTHORIZED = 'UNAUTHORIZED',
  FORBIDDEN = 'FORBIDDEN',
  
  // Method & Request Validation
  METHOD_NOT_ALLOWED = 'METHOD_NOT_ALLOWED',
  MISSING_MESSAGE = 'MISSING_MESSAGE',
  INVALID_MESSAGES = 'INVALID_MESSAGES',
  INVALID_REQUEST_FORMAT = 'INVALID_REQUEST_FORMAT',
  
  // OpenAI & Chat Service
  OPENAI_NOT_CONFIGURED = 'OPENAI_NOT_CONFIGURED',
  OPENAI_ERROR = 'OPENAI_ERROR',
  CHAT_ERROR = 'CHAT_ERROR',
  INVALID_JSON_RESPONSE = 'INVALID_JSON_RESPONSE',
  TOKEN_LIMIT_EXCEEDED = 'TOKEN_LIMIT_EXCEEDED',
  
  // Rate Limiting
  RATE_LIMIT_EXCEEDED = 'RATE_LIMIT_EXCEEDED',
  
  // File Upload & Processing
  NO_FILE = 'NO_FILE',
  FILE_TOO_LARGE = 'FILE_TOO_LARGE',
  INVALID_FILE_TYPE = 'INVALID_FILE_TYPE',
  UPLOAD_ERROR = 'UPLOAD_ERROR',
  STORAGE_ERROR = 'STORAGE_ERROR',
  PDF_VALIDATION_FAILED = 'PDF_VALIDATION_FAILED',
  INVALID_PDF = 'INVALID_PDF',
  
  // Database Operations
  DATABASE_ERROR = 'DATABASE_ERROR',
  SESSION_ERROR = 'SESSION_ERROR',
  SESSION_NOT_FOUND = 'SESSION_NOT_FOUND',
  USER_NOT_FOUND = 'USER_NOT_FOUND',
  
  // Document Processing
  DOCUMENT_CONTEXT_ERROR = 'DOCUMENT_CONTEXT_ERROR',
  SEARCH_ERROR = 'SEARCH_ERROR',
  
  // OM Analysis Specific
  OM_VALIDATION_FAILED = 'OM_VALIDATION_FAILED',
  OM_PARSING_ERROR = 'OM_PARSING_ERROR',
  OM_SCHEMA_MISMATCH = 'OM_SCHEMA_MISMATCH',
  
  // Security
  CONTENT_POLICY_VIOLATION = 'CONTENT_POLICY_VIOLATION',
  PROMPT_INJECTION_DETECTED = 'PROMPT_INJECTION_DETECTED',
  INPUT_SANITIZATION_FAILED = 'INPUT_SANITIZATION_FAILED',
  
  // Configuration & Environment
  CONFIG_ERROR = 'CONFIG_ERROR',
  ENVIRONMENT_ERROR = 'ENVIRONMENT_ERROR',
  
  // Generic Errors
  INTERNAL_ERROR = 'INTERNAL_ERROR',
  UNKNOWN_ERROR = 'UNKNOWN_ERROR',
  VALIDATION_ERROR = 'VALIDATION_ERROR'
}

/**
 * User-friendly error messages for each error code
 */
export const ERROR_MESSAGES: Record<ERROR_CODES, string> = {
  // Authentication & Authorization
  [ERROR_CODES.MISSING_TOKEN]: 'Authentication token is required',
  [ERROR_CODES.INVALID_TOKEN]: 'Invalid authentication token',
  [ERROR_CODES.UNAUTHORIZED]: 'Authentication required to access this resource',
  [ERROR_CODES.FORBIDDEN]: 'You do not have permission to access this resource',
  
  // Method & Request Validation
  [ERROR_CODES.METHOD_NOT_ALLOWED]: 'HTTP method not allowed for this endpoint',
  [ERROR_CODES.MISSING_MESSAGE]: 'Message is required',
  [ERROR_CODES.INVALID_MESSAGES]: 'Invalid messages format',
  [ERROR_CODES.INVALID_REQUEST_FORMAT]: 'Request format is invalid',
  
  // OpenAI & Chat Service
  [ERROR_CODES.OPENAI_NOT_CONFIGURED]: 'AI service is temporarily unavailable',
  [ERROR_CODES.OPENAI_ERROR]: 'AI service error occurred',
  [ERROR_CODES.CHAT_ERROR]: 'Failed to process chat request',
  [ERROR_CODES.INVALID_JSON_RESPONSE]: 'AI returned invalid response format',
  [ERROR_CODES.TOKEN_LIMIT_EXCEEDED]: 'Message too long, please shorten your request',
  
  // Rate Limiting
  [ERROR_CODES.RATE_LIMIT_EXCEEDED]: 'Too many requests, please try again later',
  
  // File Upload & Processing
  [ERROR_CODES.NO_FILE]: 'No file was uploaded',
  [ERROR_CODES.FILE_TOO_LARGE]: 'File size exceeds maximum limit',
  [ERROR_CODES.INVALID_FILE_TYPE]: 'Invalid file type, only PDF files allowed',
  [ERROR_CODES.UPLOAD_ERROR]: 'File upload failed',
  [ERROR_CODES.STORAGE_ERROR]: 'File storage error occurred',
  [ERROR_CODES.PDF_VALIDATION_FAILED]: 'PDF file validation failed',
  [ERROR_CODES.INVALID_PDF]: 'Invalid PDF file format',
  
  // Database Operations
  [ERROR_CODES.DATABASE_ERROR]: 'Database operation failed',
  [ERROR_CODES.SESSION_ERROR]: 'Session management error',
  [ERROR_CODES.SESSION_NOT_FOUND]: 'Chat session not found',
  [ERROR_CODES.USER_NOT_FOUND]: 'User account not found',
  
  // Document Processing
  [ERROR_CODES.DOCUMENT_CONTEXT_ERROR]: 'Failed to process document context',
  [ERROR_CODES.SEARCH_ERROR]: 'Document search failed',
  
  // OM Analysis Specific
  [ERROR_CODES.OM_VALIDATION_FAILED]: 'OM analysis validation failed',
  [ERROR_CODES.OM_PARSING_ERROR]: 'Failed to parse offering memorandum',
  [ERROR_CODES.OM_SCHEMA_MISMATCH]: 'OM response does not match expected format',
  
  // Security
  [ERROR_CODES.CONTENT_POLICY_VIOLATION]: 'Content violates usage policies',
  [ERROR_CODES.PROMPT_INJECTION_DETECTED]: 'Security violation detected in input',
  [ERROR_CODES.INPUT_SANITIZATION_FAILED]: 'Input validation failed',
  
  // Configuration & Environment
  [ERROR_CODES.CONFIG_ERROR]: 'Server configuration error',
  [ERROR_CODES.ENVIRONMENT_ERROR]: 'Environment setup error',
  
  // Generic Errors
  [ERROR_CODES.INTERNAL_ERROR]: 'Internal server error occurred',
  [ERROR_CODES.UNKNOWN_ERROR]: 'An unexpected error occurred',
  [ERROR_CODES.VALIDATION_ERROR]: 'Data validation failed'
};

/**
 * HTTP status codes for each error type
 */
export const ERROR_STATUS_CODES: Record<ERROR_CODES, number> = {
  // Authentication & Authorization (401, 403)
  [ERROR_CODES.MISSING_TOKEN]: 401,
  [ERROR_CODES.INVALID_TOKEN]: 401,
  [ERROR_CODES.UNAUTHORIZED]: 401,
  [ERROR_CODES.FORBIDDEN]: 403,
  
  // Method & Request Validation (400, 405)
  [ERROR_CODES.METHOD_NOT_ALLOWED]: 405,
  [ERROR_CODES.MISSING_MESSAGE]: 400,
  [ERROR_CODES.INVALID_MESSAGES]: 400,
  [ERROR_CODES.INVALID_REQUEST_FORMAT]: 400,
  
  // OpenAI & Chat Service (500, 503)
  [ERROR_CODES.OPENAI_NOT_CONFIGURED]: 503,
  [ERROR_CODES.OPENAI_ERROR]: 500,
  [ERROR_CODES.CHAT_ERROR]: 500,
  [ERROR_CODES.INVALID_JSON_RESPONSE]: 500,
  [ERROR_CODES.TOKEN_LIMIT_EXCEEDED]: 413,
  
  // Rate Limiting (429)
  [ERROR_CODES.RATE_LIMIT_EXCEEDED]: 429,
  
  // File Upload & Processing (400, 413)
  [ERROR_CODES.NO_FILE]: 400,
  [ERROR_CODES.FILE_TOO_LARGE]: 413,
  [ERROR_CODES.INVALID_FILE_TYPE]: 400,
  [ERROR_CODES.UPLOAD_ERROR]: 500,
  [ERROR_CODES.STORAGE_ERROR]: 500,
  [ERROR_CODES.PDF_VALIDATION_FAILED]: 400,
  [ERROR_CODES.INVALID_PDF]: 400,
  
  // Database Operations (500, 404)
  [ERROR_CODES.DATABASE_ERROR]: 500,
  [ERROR_CODES.SESSION_ERROR]: 500,
  [ERROR_CODES.SESSION_NOT_FOUND]: 404,
  [ERROR_CODES.USER_NOT_FOUND]: 404,
  
  // Document Processing (500)
  [ERROR_CODES.DOCUMENT_CONTEXT_ERROR]: 500,
  [ERROR_CODES.SEARCH_ERROR]: 500,
  
  // OM Analysis Specific (400, 422)
  [ERROR_CODES.OM_VALIDATION_FAILED]: 422,
  [ERROR_CODES.OM_PARSING_ERROR]: 422,
  [ERROR_CODES.OM_SCHEMA_MISMATCH]: 422,
  
  // Security (400, 403)
  [ERROR_CODES.CONTENT_POLICY_VIOLATION]: 403,
  [ERROR_CODES.PROMPT_INJECTION_DETECTED]: 400,
  [ERROR_CODES.INPUT_SANITIZATION_FAILED]: 400,
  
  // Configuration & Environment (500)
  [ERROR_CODES.CONFIG_ERROR]: 500,
  [ERROR_CODES.ENVIRONMENT_ERROR]: 500,
  
  // Generic Errors (500, 422)
  [ERROR_CODES.INTERNAL_ERROR]: 500,
  [ERROR_CODES.UNKNOWN_ERROR]: 500,
  [ERROR_CODES.VALIDATION_ERROR]: 422
};

/**
 * Typed error response interface
 */
export interface TypedApiError {
  error: string;
  code: ERROR_CODES;
  details?: string;
  timestamp?: string;
}

/**
 * Enhanced API error response helper
 * @param res - Next.js response object
 * @param code - Error code from enum
 * @param details - Optional additional details
 * @param customMessage - Optional custom message override
 */
export function createApiError(
  res: any, 
  code: ERROR_CODES, 
  details?: string,
  customMessage?: string
): void {
  const statusCode = ERROR_STATUS_CODES[code];
  const message = customMessage || ERROR_MESSAGES[code];
  
  const errorResponse: TypedApiError = {
    error: message,
    code,
    timestamp: new Date().toISOString(),
    ...(details && { details })
  };
  
  res.status(statusCode).json(errorResponse);
}

/**
 * Check if an error code indicates a client error (4xx)
 * @param code - Error code to check
 * @returns True if client error
 */
export function isClientError(code: ERROR_CODES): boolean {
  const statusCode = ERROR_STATUS_CODES[code];
  return statusCode >= 400 && statusCode < 500;
}

/**
 * Check if an error code indicates a server error (5xx)
 * @param code - Error code to check
 * @returns True if server error
 */
export function isServerError(code: ERROR_CODES): boolean {
  const statusCode = ERROR_STATUS_CODES[code];
  return statusCode >= 500;
}

/**
 * Get retry-able error codes (temporary failures)
 */
export const RETRYABLE_ERROR_CODES: Set<ERROR_CODES> = new Set([
  ERROR_CODES.RATE_LIMIT_EXCEEDED,
  ERROR_CODES.OPENAI_ERROR,
  ERROR_CODES.DATABASE_ERROR,
  ERROR_CODES.STORAGE_ERROR,
  ERROR_CODES.INTERNAL_ERROR,
  ERROR_CODES.UPLOAD_ERROR
]);

/**
 * Check if an error code indicates a retryable error
 * @param code - Error code to check
 * @returns True if error is retryable
 */
export function isRetryableError(code: ERROR_CODES): boolean {
  return RETRYABLE_ERROR_CODES.has(code);
}