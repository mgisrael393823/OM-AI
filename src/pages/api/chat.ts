import { NextApiRequest, NextApiResponse } from 'next'
import { createClient } from '@supabase/supabase-js'
import { withAuth, withRateLimit, apiError, AuthenticatedRequest } from '@/lib/auth-middleware'
import { openai, isOpenAIConfigured } from '@/lib/openai-client'
import { checkEnvironment, getConfig } from '@/lib/config'
import { logError } from '@/lib/error-logger'
import type { Database } from '@/types/database'

// Unified request type supporting both simple and complex formats
interface UnifiedChatRequest {
  // Simple format (chat-enhanced compatibility)
  message?: string
  sessionId?: string
  chat_session_id?: string // legacy field
  documentId?: string
  document_id?: string // legacy field
  
  // Complex format (chat-v2 compatibility)
  messages?: Array<{
    role: 'user' | 'assistant' | 'system'
    content: string
  }>
  documentContext?: {
    documentIds: string[]
    maxChunks?: number
    relevanceThreshold?: number
  }
  options?: {
    model?: string
    temperature?: number
    maxTokens?: number
    stream?: boolean
  }
}

// Migration helper function to normalize different request formats
function normalizeRequest(body: any): UnifiedChatRequest {
  return {
    message: body.message,
    messages: body.messages,
    sessionId: body.sessionId || body.chat_session_id,
    documentId: body.documentId || body.document_id,
    documentContext: body.documentContext,
    options: body.options || {}
  }
}

async function chatHandler(req: AuthenticatedRequest, res: NextApiResponse) {
  // Request body logging removed for production
  
  if (req.method !== 'POST') {
    return apiError(res, 405, 'Method not allowed', 'METHOD_NOT_ALLOWED')
  }

  // Validate environment at runtime (non-blocking)
  const validation = checkEnvironment()
  if (!validation.isValid) {
    console.warn("Environment validation warnings:", validation.errors)
  }

  // Check if OpenAI is properly configured
  if (!isOpenAIConfigured()) {
    return apiError(res, 503, "Chat service unavailable", "OPENAI_NOT_CONFIGURED",
      "OpenAI API key is not configured. Please contact support.")
  }

  // Normalize and validate request format
  const normalizedRequest = normalizeRequest(req.body)
  const isSimple = typeof normalizedRequest.message === 'string'

  // Validate request format
  if (isSimple && !normalizedRequest.message) {
    return apiError(res, 400, 'Message is required for simple format', 'MISSING_MESSAGE')
  }
  if (!isSimple && (!normalizedRequest.messages || !Array.isArray(normalizedRequest.messages))) {
    return apiError(res, 400, 'Messages array is required for complex format', 'INVALID_MESSAGES')
  }

  // Log deprecated endpoint usage
  const deprecatedEndpoint = req.headers['x-deprecated-endpoint'] as string
  if (deprecatedEndpoint) {
    console.warn(`Deprecated endpoint used: /api/${deprecatedEndpoint}`, {
      userId: req.user.id,
      userAgent: req.headers['user-agent'],
      timestamp: new Date().toISOString()
    })
  }

  const config = getConfig()
  const supabase = createClient<Database>(
    config.supabase.url,
    config.supabase.serviceRoleKey
  )

  // Apply rate limiting per user (simplified for now)
  try {
    await withRateLimit(req.user.id, 20, 2, async () => {
      // Rate limit: 20 requests per user, refill 2 tokens per minute
    })
  } catch (error) {
    return apiError(res, 429, "Rate limit exceeded. Please try again later.", "RATE_LIMIT_EXCEEDED")
  }

  try {
    // Handle session management
    let sessionId = normalizedRequest.sessionId
    if (isSimple && !sessionId) {
      // Auto-create session for simple format
      const { data: newSession, error: sessionError } = await supabase
        .from('chat_sessions')
        .insert({
          user_id: req.user.id,
          title: normalizedRequest.message!.slice(0, 50) + (normalizedRequest.message!.length > 50 ? '...' : ''),
          document_id: normalizedRequest.documentId || null
        })
        .select()
        .single()

      if (sessionError) {
        return apiError(res, 500, 'Failed to create chat session', 'SESSION_ERROR')
      }

      sessionId = newSession.id
    } else if (sessionId) {
      // Verify session belongs to user
      const { error: sessionVerifyError } = await supabase
        .from('chat_sessions')
        .select('id')
        .eq('id', sessionId)
        .eq('user_id', req.user.id)
        .single()

      if (sessionVerifyError) {
        return apiError(res, 404, 'Chat session not found', 'SESSION_NOT_FOUND')
      }
    }

    // Handle document context
    let contextualInformation = ""
    const documentIds = normalizedRequest.documentContext?.documentIds || 
      (normalizedRequest.documentId ? [normalizedRequest.documentId] : [])
    
    if (documentIds.length > 0) {
      try {
        const userQuery = isSimple ? normalizedRequest.message : 
          normalizedRequest.messages?.[normalizedRequest.messages.length - 1]?.content || ""
        
        // Get document chunks
        const { data: relevantChunks, error: searchError } = await supabase
          .from('document_chunks')
          .select(`
            content,
            page_number,
            chunk_type,
            documents!inner(original_filename)
          `)
          .eq('user_id', req.user.id)
          .in('document_id', documentIds)
          .textSearch('content', userQuery)
          .limit(5)

        if (!searchError && relevantChunks && relevantChunks.length > 0) {
          contextualInformation = `

DOCUMENT CONTEXT:
The following information is from the user's uploaded documents:

${relevantChunks
  .map((chunk, index) => {
    const docName = (chunk as any).documents?.original_filename ?? 'Unknown';
    return `[${index + 1}] From "${docName}" (Page ${chunk.page_number}):
${chunk.content.substring(0, 800)}${chunk.content.length > 800 ? '...' : ''}`;
  })
  .join('\n')}

Please reference this document context in your response when relevant.`
        }
      } catch (error) {
        console.error('Error retrieving document context:', error)
      }
    }

    // Build messages for OpenAI
    let messages: Array<{ role: 'user' | 'assistant' | 'system', content: string }> = []
    
    if (isSimple) {
      // For simple format, save user message and get conversation history
      if (sessionId) {
        await supabase.from('messages').insert({
          chat_session_id: sessionId,
          role: 'user',
          content: normalizedRequest.message!
        })

        // Get conversation history
        const { data: history } = await supabase
          .from('messages')
          .select('role, content')
          .eq('chat_session_id', sessionId)
          .order('created_at', { ascending: true })

        messages = (history || []) as Array<{ role: 'user' | 'assistant' | 'system', content: string }>
      } else {
        // No session, just use the current message
        messages = [{ role: 'user', content: normalizedRequest.message! }]
      }
    } else {
      // For complex format, use provided messages
      messages = normalizedRequest.messages!
    }

    const systemMessage = {
      role: "system" as const,
      content: `You are OM Intel, an advanced AI assistant specializing in commercial real estate analysis and document review. You are professional, insightful, and highly knowledgeable about:

- Commercial real estate transactions and valuations
- Property investment analysis and due diligence
- Market trends and comparative analysis
- Financial modeling and cash flow projections
- Lease agreements and property management
- Zoning, development, and regulatory matters
- Risk assessment and mitigation strategies

Your communication style is:
- Professional yet approachable
- Clear and concise
- Data-driven with actionable insights
- Focused on helping users make informed decisions

When analyzing documents or answering questions:
- Provide specific, detailed analysis
- Highlight key risks and opportunities
- Offer practical recommendations
- Ask clarifying questions when needed
- Reference relevant market standards and best practices
- When document context is provided, reference specific details from the documents
- Cite page numbers and document names when referencing uploaded content

Always maintain confidentiality and provide accurate, helpful information to support commercial real estate professionals in their decision-making process.${contextualInformation}`
    }

    // Always use SSE format for backward compatibility (since frontend expects it)
    const useSSEFormat = true
    const shouldStream = normalizedRequest.options?.stream !== false

    if (shouldStream) {
      // Set up streaming headers
      const headers: Record<string, string> = {
        'Cache-Control': 'no-cache',
        'Connection': 'keep-alive'
      }
      
      if (useSSEFormat) {
        headers['Content-Type'] = 'text/event-stream'
        headers['X-Accel-Buffering'] = 'no'
      } else {
        headers['Content-Type'] = 'text/plain; charset=utf-8'
      }
      
      if (sessionId) {
        headers['X-Chat-Session-Id'] = sessionId
      }
      
      res.writeHead(200, headers)

      // Create streaming response
      const response = await openai.chat.completions.create({
        model: normalizedRequest.options?.model || "gpt-4o",
        messages: [systemMessage, ...messages],
        temperature: normalizedRequest.options?.temperature || 0.7,
        max_tokens: normalizedRequest.options?.maxTokens || 2000,
        stream: true,
      })

      let assistantResponse = ""
      let buffer = ''
      let lastFlush = Date.now()
      const FLUSH_INTERVAL = 50
      const MIN_CHUNK_SIZE = 5
      
      const flushBuffer = () => {
        if (buffer) {
          if (useSSEFormat) {
            res.write(`data: ${JSON.stringify({ content: buffer })}\n\n`)
          } else {
            res.write(buffer)
          }
          assistantResponse += buffer
          buffer = ''
          lastFlush = Date.now()
        }
      }

      for await (const chunk of response) {
        console.log('🔄 CHUNK:', chunk.choices[0]?.delta || chunk)
        const content = chunk.choices[0]?.delta?.content || ""
        if (content) {
          buffer += content
          
          const shouldFlush = buffer.length >= MIN_CHUNK_SIZE || 
                             (Date.now() - lastFlush) >= FLUSH_INTERVAL
          
          if (shouldFlush) {
            flushBuffer()
          }
        }
      }
      
      flushBuffer()

      if (useSSEFormat) {
        res.write(`data: ${JSON.stringify({ done: true })}\n\n`)
      }

      // Save assistant response if we have a session
      if (sessionId && assistantResponse) {
        await supabase.from('messages').insert({
          chat_session_id: sessionId,
          role: 'assistant',
          content: assistantResponse
        })

        await supabase.from('chat_sessions')
          .update({ updated_at: new Date().toISOString() })
          .eq('id', sessionId)
      }

      res.end()
    } else {
      // Non-streaming response
      const response = await openai.chat.completions.create({
        model: normalizedRequest.options?.model || "gpt-4o",
        messages: [systemMessage, ...messages],
        temperature: normalizedRequest.options?.temperature || 0.7,
        max_tokens: normalizedRequest.options?.maxTokens || 2000,
        stream: false,
      })

      const assistantContent = response.choices[0]?.message?.content || ""

      // Save messages if we have a session
      if (sessionId) {
        await supabase.from('messages').insert({
          chat_session_id: sessionId,
          role: 'assistant',
          content: assistantContent
        })

        await supabase.from('chat_sessions')
          .update({ updated_at: new Date().toISOString() })
          .eq('id', sessionId)
      }

      res.status(200).json({
        id: response.id,
        content: assistantContent,
        model: response.model,
        usage: response.usage,
        sessionId
      })
    }

  } catch (error) {
    logError(error, {
      endpoint: '/api/chat',
      userId: req.user.id,
      errorType: 'CHAT_ERROR'
    })
    
    if (!res.headersSent) {
      return apiError(res, 500, "Failed to get response from AI", "OPENAI_ERROR",
        error instanceof Error ? error.message : "Unknown error")
    } else {
      if (deprecatedEndpoint === 'chat-enhanced' || deprecatedEndpoint === 'chat') {
        res.write(`data: ${JSON.stringify({ 
          error: error instanceof Error ? error.message : "Unknown error" 
        })}\n\n`)
      }
      res.end()
    }
  }
}

export default function handler(req: NextApiRequest, res: NextApiResponse) {
  return withAuth(req, res, chatHandler)
}

export { chatHandler }
