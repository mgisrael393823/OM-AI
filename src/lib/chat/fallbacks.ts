import { callOpenAIWithFallback, type CallOptions } from '@/lib/services/openai/client-wrapper'

export function callWithFallback(
  options: CallOptions & { requestId: string; signal?: AbortSignal },
  customModel?: string
): Promise<any> {
  return callOpenAIWithFallback(options, customModel)
}

// TODO: expand cascade/text fallback utilities