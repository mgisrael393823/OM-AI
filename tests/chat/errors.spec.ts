/** @jest-environment node */
import { createResponse, createRequest } from 'node-mocks-http'
import { jsonError } from '@/lib/chat/errors'

describe('chat errors', () => {
  it('produces legacy error shape by default', () => {
    const req = createRequest()
    const res = createResponse()
    jsonError(res as any, 400, 'TEST_CODE', 'Test message', 'req-123', req as any)
    expect(res._getJSONData()).toEqual({
      error: 'Test message',
      code: 'TEST_CODE',
      requestId: 'req-123'
    })
  })

  it('produces structured error shape when requested via header', () => {
    const req = createRequest({
      headers: {
        'x-error-format': 'structured'
      }
    })
    const res = createResponse()
    jsonError(res as any, 400, 'TEST_CODE', 'Test message', 'req-123', req as any)
    expect(res._getJSONData()).toMatchSnapshot()
  })

  it('produces structured error shape when requested via query param', () => {
    const req = createRequest({
      query: {
        errorFormat: 'structured'
      }
    })
    const res = createResponse()
    jsonError(res as any, 400, 'TEST_CODE', 'Test message', 'req-123', req as any)
    expect(res._getJSONData()).toMatchSnapshot()
  })

  it('defaults to legacy format without request object', () => {
    const res = createResponse()
    jsonError(res as any, 400, 'TEST_CODE', 'Test message', 'req-123')
    expect(res._getJSONData()).toEqual({
      error: 'Test message',
      code: 'TEST_CODE',
      requestId: 'req-123'
    })
  })
})