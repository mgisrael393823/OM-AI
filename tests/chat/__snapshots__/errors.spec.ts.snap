// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat errors produces structured error shape when requested via header 1`] = `
{
  "error": {
    "code": "TEST_CODE",
    "message": "Test message",
    "requestId": "req-123",
    "type": "api_error",
  },
}
`;

exports[`chat errors produces structured error shape when requested via query param 1`] = `
{
  "error": {
    "code": "TEST_CODE",
    "message": "Test message",
    "requestId": "req-123",
    "type": "api_error",
  },
}
`;